File: benchmarks/tsb/bench_pct_change_na.ts
/**
 * Benchmark: pctChangeSeries / pctChangeDataFrame — percent change computations.
 * Outputs JSON: {"function": "pct_change_na", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { Series, DataFrame, pctChangeSeries, pctChangeDataFrame } from "../../src/index.ts";

const SIZE = 100_000;
const ROWS = 10_000;
const WARMUP = 5;
const ITERATIONS = 50;

const s = new Series({ data: Array.from({ length: SIZE }, (_, i) => 100 + Math.sin(i / 100)) });
const df = DataFrame.fromColumns({
  price: Array.from({ length: ROWS }, (_, i) => 100 + i * 0.01),
  volume: Array.from({ length: ROWS }, (_, i) => 1000 + (i % 100) * 10),
  ratio: Array.from({ length: ROWS }, (_, i) => 0.5 + Math.cos(i / 1000)),
});

for (let i = 0; i < WARMUP; i++) {
  pctChangeSeries(s);
  pctChangeSeries(s, { periods: 5 });
  pctChangeDataFrame(df);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  pctChangeSeries(s);
  pctChangeSeries(s, { periods: 5 });
  pctChangeDataFrame(df);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "pct_change_na",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);

File: benchmarks/pandas/bench_pct_change_na.py
"""
Benchmark: Series.pct_change() / DataFrame.pct_change() — percent change computations.
Outputs JSON: {"function": "pct_change_na", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 100_000
ROWS = 10_000
WARMUP = 5
ITERATIONS = 50

s = pd.Series(100 + np.sin(np.arange(SIZE) / 100))
df = pd.DataFrame({
    "price": 100 + np.arange(ROWS) * 0.01,
    "volume": 1000 + (np.arange(ROWS) % 100) * 10,
    "ratio": 0.5 + np.cos(np.arange(ROWS) / 1000),
})

for _ in range(WARMUP):
    s.pct_change()
    s.pct_change(periods=5)
    df.pct_change()

start = time.perf_counter()
for _ in range(ITERATIONS):
    s.pct_change()
    s.pct_change(periods=5)
    df.pct_change()
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "pct_change_na",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))
